compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/extended/VariableAssignmentChecker.kt

```diff
@@ -0,0 +1,125 @@
+                    reporter.report(lastDestructuringSource, FirErrors.CAN_BE_VAL)
+                    lastDestructuringSource = null
+                } else if (!canBeVal(symbol, value)) {
+                    destructuringCanBeVal = false
+                }
+                lastDestructuredVariables--
+            } else if (canBeVal(symbol, value)) {
+                reporter.report(source, FirErrors.CAN_BE_VAL)
+            }
+        }
+    }
+
+    private fun canBeVal(symbol: FirPropertySymbol, value: AssignmentsCount) = value == AssignmentsCount.ONE && symbol.fir.isVar
+
+    private class UninitializedPropertyReporter(
+        val data: Map<CFGNode<*>, PropertyInitializationInfo>,
+        val localProperties: Set<FirPropertySymbol>,
+        val unprocessedProperties: MutableSet<FirPropertySymbol>,
+        val propertiesCharacteristics: MutableMap<FirPropertySymbol, AssignmentsCount>
+    ) : ControlFlowGraphVisitorVoid() {
+        override fun visitNode(node: CFGNode<*>) {}
+
+        override fun visitVariableAssignmentNode(node: VariableAssignmentNode) {
+            val symbol = (node.fir.calleeReference as? FirResolvedNamedReference)?.resolvedSymbol as? FirPropertySymbol ?: return
+            if (symbol !in localProperties) return
+            unprocessedProperties.remove(symbol)
+
+            val currentCharacteristic = propertiesCharacteristics.getOrDefault(symbol, AssignmentsCount.ZERO)
+
+            when (data.getValue(node)[symbol] ?: EventOccurrencesRange.ZERO) {
+                EventOccurrencesRange.AT_MOST_ONCE -> {
+                    propertiesCharacteristics[symbol] = maxOf(currentCharacteristic, AssignmentsCount.ONE)
+                }
+                EventOccurrencesRange.EXACTLY_ONCE -> {
+                    propertiesCharacteristics[symbol] = maxOf(currentCharacteristic, AssignmentsCount.ONE)
+                }
+                EventOccurrencesRange.AT_LEAST_ONCE, EventOccurrencesRange.MORE_THAN_ONCE -> {
+                    propertiesCharacteristics[symbol] = AssignmentsCount.MANY
+                }
+                else -> {
+                    propertiesCharacteristics[symbol] = maxOf(currentCharacteristic, AssignmentsCount.ZERO)
+                }
+            }
+        }
+
+        override fun visitVariableDeclarationNode(node: VariableDeclarationNode) {
+            val symbol = node.fir.symbol
+            if (node.fir.initializer == null && node.fir.delegate == null) {
+                unprocessedProperties.add(symbol)
+            } else {
+                propertiesCharacteristics[symbol] = AssignmentsCount.ONE
+            }
+        }
+    }
+
+    private val FirPropertySymbol.getValOrVarSource
+        get() = (fir.psi as? KtProperty)?.valOrVarKeyword?.toFirPsiSourceElement()
+            ?: fir.psi?.firstChild?.toFirPsiSourceElement()
+            ?: fir.source
+
+    enum class AssignmentsCount {
```

Comment:
You don't need `AssignmentsCount` since you have extended `EventOccurrencesRange`